server/src/main/java/org/opensearch/tasks/ResourceUsageInfo.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.tasks;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.EnumMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Thread resource usage information for particular resource stats type.
+ * <p>
+ * It captures the resource usage information like memory, CPU about a particular execution of thread
+ * for a specific stats type.
+ */
+public class ResourceUsageInfo {
+    private static final Logger logger = LogManager.getLogger(ResourceUsageInfo.class);
+    private final EnumMap<ResourceStats, ResourceStatsInfo> statsInfo = new EnumMap<>(ResourceStats.class);
+
+    public ResourceUsageInfo(ResourceUsageMetric... resourceUsageMetrics) {
+        for (ResourceUsageMetric resourceUsageMetric : resourceUsageMetrics) {
+            this.statsInfo.put(resourceUsageMetric.getStats(), new ResourceStatsInfo(resourceUsageMetric.getValue()));
+        }
+    }
+
+    public void recordResourceUsageMetrics(ResourceUsageMetric... resourceUsageMetrics) {
+        for (ResourceUsageMetric resourceUsageMetric : resourceUsageMetrics) {
+            final ResourceStatsInfo resourceStatsInfo = statsInfo.get(resourceUsageMetric.getStats());
+            if (resourceStatsInfo != null) {
+                updateResourceUsageInfo(resourceStatsInfo, resourceUsageMetric);
+            } else {
+                throw new IllegalStateException(
+                    "cannot update ["
+                        + resourceUsageMetric.getStats().toString()
+                        + "] entry as its not present current_stats_info:"
+                        + statsInfo
+                );
+            }
+        }
+    }
+
+    private void updateResourceUsageInfo(ResourceStatsInfo resourceStatsInfo, ResourceUsageMetric resourceUsageMetric) {
+        long currentEndValue;
+        long newEndValue;
+        do {
+            currentEndValue = resourceStatsInfo.endValue.get();
+            newEndValue = resourceUsageMetric.getValue();
+            if (currentEndValue > newEndValue) {
```

Comment:
I'm probably being too pedantic here, but this comparison makes an assumption that the usage metric will always go up and you would always want to drop the lower value in case of a conflict. That's true now but is it truly inherent for all possible values that may be tracked here? Maybe the answer is yes...something to think about. Not a blocker.

Reply:
All the resource stats are thread specific and expected to be fetched from ```ThreadMxBean```, which returns with a non-decreasing value all the time. ```update``` thread or ```stop``` thread calls the ```ThreadMxBean```, is returned with non-decreasing value. 

_Case1:_
When ```update``` thread calls ```ThreadMxBean``` before``` stop``` thread, the update will have smaller value and stop will have the higher value. If updates from both these threads occur concurrently the value from stop thread will persist (as its higher).
_Case2:_ 
When ```update``` calls ```ThreadMxBean``` after ``` stop```, the update will have higher value. We are okay with storing the value from update thread over the value reported from stop thread as the thread is still tied to the task. This helps us avoid lock mechanism between stop and update. 
_Case3:_
When update calls with previously cached value. The stale cached value will be lower than value stop thread will report.